crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -1154,6 +1157,80 @@ fn handle_attribute_comment<'a>(
     }
 }
 
+/// Assign comments between `if` and `test` and `else` and `orelse` as leading to the respective
+/// node.
+///
+/// ```python
+/// x = (
+///     "a"
+///     if # leading comment of `True`
+///     True
+///     else # leading comment of `"b"`
+///     "b"
+/// )
+/// ```
+///
+/// This placement ensures comments remain in their previous order. This an edge case that only
+/// happens if the comments are in a weird position but it also doesn't hurt handling it.
+fn handle_expr_if_comment<'a>(
+    comment: DecoratedComment<'a>,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    let Some(expr_if) = comment.enclosing_node().expr_if_exp() else {
+        return CommentPlacement::Default(comment);
+    };
+    let ExprIfExp {
+        range: _,
+        test,
+        body,
+        orelse,
+    } = expr_if;
+
+    // Find the if and the else
+    let if_token =
+        find_only_token_str_in_range(TextRange::new(body.end(), test.start()), locator, "if");
+    let else_token =
+        find_only_token_str_in_range(TextRange::new(test.end(), orelse.start()), locator, "else");
+
+    if if_token.range.start() < comment.slice().start()
+        && comment.slice().start() < test.start()
+        && comment.line_position().is_end_of_line()
+    {
+        return CommentPlacement::leading(test.as_ref().into(), comment);
+    }
+
+    if else_token.range.start() < comment.slice().start()
+        && comment.slice().start() < orelse.start()
+        && comment.line_position().is_end_of_line()
+    {
+        return CommentPlacement::leading(orelse.as_ref().into(), comment);
+    }
+
+    CommentPlacement::Default(comment)
+}
+
+/// Looks for a multi char token in the range that contains no other tokens. `SimpleTokenizer` only
+/// works with single char tokens so we check that we have the right token by string comparison.
```

Comment:
It should be possible to add support for `if` and `else`. Comments are already `non-single` token strings. 
The backwards lexing is a bit more annoying but should be doable as well.

Reply:
i didn't want to change the single-char-ness of the tokenizer just for this PR, but if you want i can do the refactoring in another PR